readme.txt
=== ProjectManager ===
Contributors: Kolja Schleich
Donate link: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=2329191
Tags: datamanager, CMS, Content Management System
Requires at least: 2.7
Tested up to: 2.7.1
Stable tag: 2.8.2

This plugin can be used to manage any number of projects with recurrent datasets (e.g. portrait system, dvd collection)

== Description ==

This plugin is a datamanager for any recurrent datasets. It can be used to manage and list a DVD collection, to to present portraits (e.g. athlets of a team), simple tabular calendar or anything you can think of. Below is a least of features

**Features**

* add as many different projects as you want to
* widget for any project, controlled via admin panel
* adding of form fields (text, textfield, e-mail, date, url, selection, checkbox and radio list) for each project independently
* simple search of any form field and category names
* template system to easily customize frontend display
* Ajax enabled editing of datasets
* easy adding of shortcodes via TinyMCE Button
* change colorscheme for output tables via admin panel
* dataset sorting by any form field
* import and export of datasets from/to CSV file
* hook projects into user profile
* manual drag & drop sorting of datasets

See [Usage](http://wordpress.org/extend/plugins/projectmanager/other_notes/) for details on shortcodes and the template system.

Due to the growing popularity of my plugins I have launched a [website](http://kolja.galerie-neander.de/)!

[ChangeLog](http://svn.wp-plugins.org/projectmanager/trunk/changelog.txt)

== Installation ==

To install the plugin to the following steps

1. Unzip the zip-file and upload the content to your Wordpress Plugin directory.
2. Activiate the plugin via the admin plugin page.


== Screenshots ==
1. Project Overview Page
2. Settings page
3. Add different form fields dynamically
4. Easy adding of new datasets
5. Widget control panel
6. Easy adding of shortcode tags via TinyMCE Button


== Usage ==

= Shortcodes =
You can display all datasets of one project with the following code

`[project id=ID template=table|gallery cat_id=ID orderby=name|id|order|formfields-ID order=ASC|DESC single=true|false selections=true|false]`


The following list gives a short description of each attribute:

* id: the ID of the project to display
* template: the template to load. Default templates are *table* and *gallery* (default: table)
* cat_id: ID of category to display datasets of (optional, default: show datasets of all categories)
* orderby: field to order datasets by. Either *name*, *id*, *order* or *formfields-ID* where ID is the formfield ID (optional, default: name)
* order: direction of dataset ordering, either *ASC* or *DESC* (optional, default: ASC)
* single: switch to add link to single dataset page (optional, default: 'true')
* selections: switch to control display of selections navbar for categories and dataset ordering (optional, default: 'true')

ProjectManager supports templates, similar to [NextGgen Gallery](http://wordpress.org/extend/plugins/nextgen-gallery/). It comes with two default templates *table* and *gallery* which are located in the *view/table.php* and *view/gallery.php*. *table* displays the datasets in a simple tabular output whereas *gallery* shows them in a gallery only with the images on the main page. You can design your own templates and place them in

`yourthemedirectory/projectmanager/`


A more detailed description of the template system can be found below.

Further single datasets can be displayed directly with the following code

`[dataset id=x]`


where x is the dataset ID.

Finally ProjectManager comes with a simple search for specific datasets, also form fields or categories. To display the search form use the following code

`[project_search project_id=x template=compact|extend]`


where x is again the project ID and template can be either *compact* or *extend*.


= Templates =
You can customize the frontend output via templates. Templates shipped with the plugin are located in the *view* subdirectory. You can store your own templates in

`yourthemedirectory/projectmanager/`


You can either copy and edit the default templates or also create your own templates. To use the templates you need to put in the template tag the filename of it without extension. For example you have created a template

`yourthemedirectory/projectmanager/gallery-custom.php`


you can use this template to display with the following code

`[project id=x template=**gallery-custom** cat_id=y]`


= Extend User Profile =

Projects can be linked with the user profile. Any usr with capability *projectmanager_user* can use this feature. If the default role has this capability users get added to each project that has the hook activated and can edit data through their profile. To subsequently add a Wordpress user to a project, use the *Add WP User* Link in the dataset form, located next to the name field. The added dataset will be linked to this user.


= Custom icons for Admin Menu =
If you want to use custom icons for the admin menu put them in

`yourthemedirectory/projectmanager/icons/`

= Access control =

Since Version 2.5 ProjectManager has fine grained access control: 

* edit_projects: add and edit projects
* delete_projects:  delete projects
* projectmanager_settings: grants access to global settings page
* edit_formfields: grant editing of formfields
* edit_projects_settings: grant access to settings page of individual projects
* import_datasets: grant import/export of datasets
* edit_datasets: add datasets and edit own datasets
* edit_other_datasets: add and edit any datasets (needs also edit_datasets)
* delete_datasets: delete own datasets
* delete_other_datasets: delete any dataset (needs also delete_datasets)
* view_projects: grant read access to projects and dataset overview
* projectmanager_user: grants permission to use profile hook feature


= Notes on Formfields =

Currently there are two formfields that require a short statement: *numeric* and *currency*. The display of these two types varies among different countries. If the PHP Internationalization Functions (http://us3.php.net/manual/en//book.intl.php), requires at least PHP 5.2 (recommended is 5.2.4+), are installed you are all set. Otherwise there are two filters that can be used to manually control the format

* projectmanager_numeric
* projectmanager_currency

Note that these filters are for some reasons not applied if you use the AJAX edit. After reloading the page the display is correct.

== Credits ==
The ProjectManager menue icons and TinyMCE Button are taken from the Fugue Icons of http://www.pinvoke.com/.

== Changelog ==

= 2.8.2 =
* BUGFIX: removed NumberFormatter from lib/ajax.php

= 2.8.1 =
* NEW: template tag for widget
* REMOVED: NumberFormatter caused syntax error for some webhosts

= 2.8 =
* NEW: widget with 2.8 API
* NEW: template tags

= 2.7.3 =
* BUGFIX: unable to delete all formfields

= 2.7.2 =
* BUGFIX: table template

= 2.7.1 =
* BUGFIX: parse error in lib/core.php line 713

= 2.7 =
* NEW: image upload formfield type
* NEW: video field with upload
* NEW: internal link to other project
* BUGFIX: dataset ordering

= 2.6 =
* NEW: shortcode attribute to limit number of results
* NEW: shortcode attribute to get datasets with specific meta key and meta value
* NEW: shortcode attribute to get random datasets with limit of results

= 2.5.8 =
* BUGIFX: call-time pass-by-reference deprecated

= 2.5.7 =
* BUGFIX: parse error

= 2.5.6 =
* CHANGED: moved AJAX functions to separate class in lib/ajax.php
* BUGFIX: chmod of image uploads
* BUGFIX: access control for editing datasets

= 2.5.5 =
* BUGFIX: access to dataset editing page with only cap 'projectmanager_user'
* BUGFIX: profile update

= 2.5.4 =
* BUGFIX: access control for adding/editing of datasets

= 2.5.3 =
* BUGFIX: access control for adding/editing of datasets
* code cleaning

= 2.5.2 =
* BUGFIX: profile hook

= 2.5.1 =
* BUGFIX: check for double datasets

= 2.5 =
* NEW: fine grained access control
* NEW: thickbox box for adding WP Users
* NEW: add dataset upon user registration if user has capablity
* NEW: profile Hook for any number of projects
* CHANGED: only users with matching datasets are shown form in profile

= 2.4.7 =
* NEW: display ID of dataset in admin overview

= 2.4.6 =
* BUGFIX: MySQL string to get datasets of certain category

= 2.4.5 =
* BUGFIX: options

= 2.4.4 =
* CHANGED: moved project settings into project table

= 2.4.3 =
* BUGFIX: display of admin menu icons
* CHANGED: moved main ProjectManager menu back down

= 2.4.2 =
* NEW: table template with tiny image
* BUGFIX: unsetting of widget option if deleted
* UPDATED: german translation with slideshow effects

= 2.4.1 =
* NEW: numeric and currency formfield types (require NumberFormatter or WP filter)
* NEW: filters projectmanager_numeric and projectmanager_currency for display
* BUGFIX: popup window size for newly added formfield options
* BUGFIX: image upload in user profile
* BUGFIX: single view in gallery template
* CHANGED: took out separate option for displaying image form in profile
* CHANGED: default limit on datasets of 15, use "NaN" for no limit

= 2.4 =
* NEW: custom icon directory in THEMEDIR/projectmanager/icons
* BUGFIX: dataset ordering is overriden by dropdown selection
* BUGFIX: IE Slideshow Bug fixed with new jQuery Cycle Plugin
* BUGFIX: Popup Window for Formfield Options
* BUGFIX: TinyMCE Button for searchform
* BUGFIX: total number of datasets in search
* CHANGED: use function add_object_page for menu if present
* CHANGED: use of jQuery Cycle Plugin for Slideshow

= 2.3 =
* NEW: URL and E-Mail with alternative text in format "www.example.com|Example Website"
* NEW: multiple widget support
* NEW: file-upload form field without AJAX uploading
* BUGFIX: AJAX editing if datafield value is empty
* BUGFIX: manual ordering of datasets
* BUGFIX: options link for select, checkbox and radio form types
* BUGFIX: save options for formfields
* CHANGED: apply stripslashes in function getDatasetMeta instead of getDatasetMetaData

= 2.2 =
* NEW: set file permission for image uploads
* NEW: image as new formfield type. Image needds to be uploaded already
* BUGFIX: AJAX editing of datasets
* BUGFIX: single and double quotes in names, formfields and formfield labels

= 2.1 =
* NEW: filter for formfields to enable bridging

= 2.0 =
* NEW: manual drag & drop sorting of datasets (ajax saving, but not updating tr class)
* NEW: set order of datasets in option
* NEW: more shortcode atts for project display
* NEW: delete image folder upon plugin uninstallation
* BUGFIX: dataset division by zero
* CHANGED: save image name in database also if file already exists

= 1.9 =
* NEW: show formfield ID on admin page
* NEW: shortcode supports ordering of datasets including TinyMCE Button
* BUGFIX: project deletion
* CHANGED: varchar lengths for project title, dataset name and image

= 1.8.2 =
* NEW: remove line breaks from form field values upon export
* BUGFIX: image upload and thumbnail creation

= 1.8.1 =
* BUGFIX: shortcode display of single dataset

= 1.8 =
* NEW: dataset ordering selection in frontend
* NEW: hook one project into profile
* NEW: set custom menu icon for each project
* NEW: shortcode to display single dataset
* NEW: template engine for frontend display
* CHANGED: new menu icons
* CHANGED: major restructuring of plugin
* CHANGED: shortcodes
* DELETED: Ajax adding of category dropdown in TinyMCE Button. Using text field for direct input of Category ID.

= 1.7 =
* NEW: sorting of datasets by form fields
* NEW: import/export of datasets from/to CSV file
* BUGFIX: form field saving

= 1.6.2 =
* CHANGED: moved image directory to wp-content/uploads

= 1.6.1 =
* BUGFIX: pretty permalinks

= 1.6 =
* NEW: Slideshow Widget
* CHANGED: usability enhancements

= 1.5 =
* NEW: coupled datasets to user id who entered them.

= 1.4 =
* NEW: selection, checkbox list, radio list form field types

= 1.3 =
* NEW: support for multiple categorization
* NEW: customization of dataset output via wordpress hooks
* NEW: search for category names (comma separated list of cat names)
* NEW: set colorschemes of tables

= 1.2.3 =
* NEW: option to add direct link to project in navigation panel
* BUGFIX: adding of new project failed

= 1.2.2 =
* BUGFIX: upgrade bug

= 1.2.1 =
* BUGFIX: database collation

= 1.2 =
* NEW: Ajax editing of datasets
* NEW: display of specific group only

= 1.1 =
* NEW: full control of display via shortcodes
* NEW: TinyMCE Button for better usability
* CHANGED: major restructuring of plugin

= 1.0 =
* initial release


changelog.txt
ProjectManager Plugin
by Kolja Schleich <kolja.schleich@googlemail.com>

V2.7.3 - July 20, 2009
- BUGFIX: unable to delete all formfields

V2.7.2 - July 14, 2009
- BUGFIX: table template

V2.7.1 - July 2, 2009
- BUGFIX: parse error in lib/core.php line 713

V2.7 - July 1, 2009
- NEW: image upload formfield type
- NEW: video field with upload
- NEW: internal link to other project
- BUGFIX: dataset ordering

V2.6 - June 20, 2009
- NEW: shortcode attribute to limit number of results
- NEW: shortcode attribute to get datasets with specific meta key and meta
value
- NEW: shortcode attribute to get random datasets with limit of results

V2.5.8 - June 8, 2009
- BUGIFX: call-time pass-by-reference deprecated

V2.5.7 - June 6, 2009
- BUGFIX: parse error

V2.5.6 - June 5, 2009
- CHANGED: moved AJAX functions to separate class in lib/ajax.php
- BUGFIX: chmod of image uploads
- BUGFIX: access control for editing datasets

V2.5.5 - June 4, 2009
- BUGFIX: access to dataset editing page with only cap 'projectmanager_user'
- BUGFIX: profile update

V2.5.4 - June 3, 2009
- BUGFIX: access control for adding/editing of datasets

V2.5.3 - June 3, 2009
- BUGFIX: access control for adding/editing of datasets
- code cleaning

V2.5.2 - June 3, 2009
- BUGFIX: profile hook

V2.5.1 - June 3, 2009
- BUGFIX: check for double datasets

V2.5 - June 2 , 2009
- NEW: fine grained access control
- NEW: thickbox box for adding WP Users
- NEW: add dataset upon user registration if user has capablity
- NEW: profile Hook for any number of projects
- CHANGED: only users with matching datasets are shown form in profile

V2.4.7 - May 30, 2009
- NEW: display ID of dataset in admin overview

V2.4.6 - May 30, 2009
- BUGFIX: MySQL string to get datasets of certain category

V2.4.5 - May 29, 2009
- BUGFIX: options

V2.4.4 - May 29, 2009
- CHANGED: moved project settings into project table

V2.4.3 - May 27, 2009
- BUGFIX: display of admin menu icons
- CHANGED: moved main ProjectManager menu back down

V2.4.2 - May 26, 2009
- NEW: table template with tiny image
- BUGFIX: unsetting of widget option if deleted
- UPDATED: german translation with slideshow effects

V2.4.1 - May 25, 2009
- NEW: numeric and currency formfield types (require NumberFormatter or WP
filter)
- NEW: filters projectmanager_numeric and projectmanager_currency for display
- BUGFIX: popup window size for newly added formfield options
- BUGFIX: image upload in user profile
- BUGFIX: single view in gallery template
- CHANGED: took out separate option for displaying image form in
profile
- CHANGED: default limit on datasets of 15, use "NaN" for no limit

V2.4 - May 15, 2009
- NEW: custom icon directory in THEMEDIR/projectmanager/icons
- BUGFIX: dataset ordering is overriden by dropdown selection
- BUGFIX: IE Slideshow Bug fixed with new jQuery Cycle Plugin
- BUGFIX: Popup Window for Formfield Options
- BUGFIX: TinyMCE Button for searchform
- BUGFIX: total number of datasets in search
- CHANGED: use function add_object_page for menu if present
- CHANGED: use of jQuery Cycle Plugin for Slideshow

V2.3 - April 14, 2009
- NEW: URL and E-Mail with alternative text in format "www.example.com|Example Website"
- NEW: multiple widget support
- NEW: file-upload form field without AJAX uploading
- BUGFIX: AJAX editing if datafield value is empty
- BUGFIX: manual ordering of datasets
- BUGFIX: options link for select, checkbox and radio form types
- BUGFIX: save options for formfields
- CHANGED: apply stripslashes in function getDatasetMeta instead of getDatasetMetaData

V2.2 - March 11, 2009
- NEW: set file permission for image uploads
- NEW: image as new formfield type. Image needds to be uploaded already
- BUGFIX: AJAX editing of datasets
- BUGFIX: single and double quotes in names, formfields and formfield labels

V2.1 - February 27, 2009
- NEW: filter for formfields to enable bridging

V2.0 - February 21, 2009
- NEW: manual drag & drop sorting of datasets (ajax saving, but not updating tr class)
- NEW: set order of datasets in option
- NEW: more shortcode atts for project display
- NEW: delete image folder upon plugin uninstallation
- BUGFIX: dataset division by zero
- CHANGED: save image name in database also if file already exists

V1.9 - February 12, 2009
- NEW: show formfield ID on admin page
- NEW: shortcode supports ordering of datasets including TinyMCE Button
- BUGFIX: project deletion
- CHANGED: varchar lengths for project title, dataset name and image

V1.8.2 - February 9, 2009
- NEW: remove line breaks from form field values upon export
- BUGFIX: image upload and thumbnail creation

V1.8.1 - February 7, 2009
- BUGFIX: shortcode display of single dataset

V1.8 - February 7, 2009
- NEW: dataset ordering selection in frontend
- NEW: hook one project into profile
- NEW: set custom menu icon for each project
- NEW: shortcode to display single dataset
- NEW: template engine for frontend display
- CHANGED: new menu icons
- CHANGED: major restructuring of plugin
- CHANGED: shortcodes
- DELETED: Ajax adding of category dropdown in TinyMCE Button. Using text field for direct input of Category ID.

V1.7 - January 23, 2009
- NEW: sorting of datasets by form fields
- NEW: import/export of datasets from/to CSV file
- BUGFIX: form field saving

V1.6.2 - January 21, 2009
- CHANGED: moved image directory to wp-content/uploads

V1.6.1 - January 18, 2009
- BUGFIX: pretty permalinks

V1.6 - January 18, 2009
- NEW: Slideshow Widget
- CHANGED: usability enhancements

V1.5 - January 5, 2009
- NEW: coupled datasets to user id who entered them.

V1.4 - December 19, 2008
- NEW: selection, checkbox list, radio list form field types

V1.3 - December 1, 2008
- NEW: support for multiple categorization
- NEW: customization of dataset output via wordpress hooks
- NEW: search for category names (comma separated list of cat names)
- NEW: set colorschemes of tables

V1.2.3 - November 24, 2008
- NEW: option to add direct link to project in navigation panel
- BUGFIX: adding of new project failed


lib/ajax.php
<?php
/**
 * AJAX class for the WordPress plugin ProjectManager
 * 
 * @author 	Kolja Schleich
 * @package	ProjectManager
 * @copyright 	Copyright 2008-2009
*/
class ProjectManagerAJAX
{
	/**
	 * add ajax actions
	 *
	 * @param none
	 * @return void
	 */
	function __construct()
	{
		// Admin AJAX
		if ( is_admin() ) {
			add_action( 'wp_ajax_projectmanager_save_name', array(&$this, 'saveName') );
			add_action( 'wp_ajax_projectmanager_save_categories', array(&$this, 'saveCategories') );
			add_action( 'wp_ajax_projectmanager_save_form_field_data', array(&$this, 'saveFormFieldData') );
			add_action( 'wp_ajax_projectmanager_save_form_field_options', array(&$this, 'saveFormFieldOptions') );
			add_action( 'wp_ajax_projectmanager_save_dataset_order', array(&$this, 'saveDatasetOrder') );
			add_action( 'wp_ajax_projectmanager_ajax_delete_file', array(&$this, 'deleteFile') ); 
			add_action( 'wp_ajax_projectmanager_insert_wp_user', array(&$this, 'insertWpUser') );

			add_action( 'wp_ajax_projectmanager_toggle_formfield_options', array(&$this, 'toggleFormfieldOptions') );
			add_action( 'wp_ajax_projectmanager_get_cat_dropdown', array(&$this, 'getCategoryDropdown') );
			add_action( 'wp_ajax_projectmanager_save_project_link', array(&$this, 'saveProjectLink') );
		}
	}
	function ProjectManagerAJAX()
	{
		$this->__construct();
	}


	/**
	 * SACK response to delete file
	 * 
	 * @since 1.4
	 */
	function deleteFile() {
		$file = $_POST['file'];
		@unlink($file);
		die();
	}


	/**
	 * SACK response for saving form field options
	 *
	 * @since 1.3
	 */
	function saveFormFieldOptions() {
		$options = get_option('projectmanager');
	
		$form_id = $_POST['form_id'];
		$form_options = substr($_POST['options'], 0, -1);
		$form_options = explode("|", $form_options);
		
		$options['form_field_options'][$form_id] = $form_options;
		update_option('projectmanager', $options);
	
		die("ProjectManager.reInit();");
	}

	
	/**
	 * save Project Link
	 *
	 * @since 2.7
	 */
	function saveProjectLink()
	{
		$project_id = (int)$_POST['project_id'];
		$formfield_id = (int)$_POST['formfield_id'];

		$options = get_option('projectmanager');
		$options['form_field_options'][$formfield_id] = $project_id;

		update_option('projectmanager', $options);
	
		die("ProjectManager.reInit();");
	}


	/**
	 * SACK response function for saving dataset name
	 *
	 * @since 1.2
	 */
	function saveName() {
		global $wpdb;
	
		$dataset_id = intval($_POST['dataset_id']);
		$new_name = $_POST['new_name'];
	
		/*if (get_magic_quotes_gpc())
			$new_name = stripslashes_deep($new_name);*/
		
		$wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->projectmanager_dataset} SET `name` = '%s' WHERE `id` = '%d'", $new_name, $dataset_id ) );

		die( "ProjectManager.reInit();jQuery('span#dataset_name_text" . $dataset_id . "').fadeOut('fast', function() {
			jQuery('a#thickboxlink_name" . $dataset_id . "').show();
			jQuery('span#dataset_name_text" . $dataset_id . "').html('" . addslashes_gpc( $new_name ) . "').fadeIn('fast');
			ProjectManager.doneLoading('loading_name_".$dataset_id."');
		});");
	}


	/**
	 * SACK response function for saving group
	 *
	 * @since 1.2
	 */
	function saveCategories() {
		global $wpdb, $projectmanager;
	
		$dataset_id = intval($_POST['dataset_id']);
		$new_cats = explode(",",substr($_POST['cat_ids'],0,-1));
	
		if ( count($new_cats) > 0 ) {
			$cat_name = $projectmanager->getSelectedCategoryTitles($new_cats);
		} else {
			$cat_name = __('None', 'projectmanager');
		}

		$wpdb->query( $wpdb->prepare ( "UPDATE {$wpdb->projectmanager_dataset} SET `cat_ids` = %s WHERE `id` = %d", maybe_serialize($new_cats), $dataset_id ) );

		die( "ProjectManager.reInit();jQuery('span#dataset_category_text" . $dataset_id . "').fadeOut('fast', function() {
			jQuery('a#thickboxlink_category" . $dataset_id . "').show();
			jQuery('span#dataset_category_text" . $dataset_id . "').html('" . $cat_name . "').fadeIn('fast');
			ProjectManager.doneLoading('loading_category_".$dataset_id."');
		});");
	}


	/**
	 * SACK response function to save any dynamic form field
	 *
	 * @since 1.2
	 */
	function saveFormFieldData() {
		global $wpdb, $projectmanager, $projectmanager_loader;
	
		$dataset_id = intval($_POST['dataset_id']);
		$formfield_type = $_POST['formfield_type'];
		$meta_id = intval($_POST['formfield_id']);
		$new_value = $_POST['new_value'];

		// Textarea
		if ( 'textfield' == $formfield_type )
			$new_value = str_replace('\n', "\n", $new_value);
		// Checkbox List
		if ( 'checkbox' == $formfield_type || 'project' == $formfield_type ) {
			$new_value = substr($new_value,0,-1);
			$new_value = explode(",",$new_value);
		}

		$count = $wpdb->get_var( "SELECT COUNT(ID) FROM {$wpdb->projectmanager_datasetmeta} WHERE `dataset_id` = '".$dataset_id."' AND `form_id` = '".$meta_id."'" );
		if ( !empty($count) )
			$wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->projectmanager_datasetmeta} SET `value` = '%s' WHERE `dataset_id` = '%d' AND `form_id` = '%d'", maybe_serialize($new_value), $dataset_id, $meta_id ) );
		else
			$wpdb->query( $wpdb->prepare( "INSERT INTO {$wpdb->projectmanager_datasetmeta} (form_id, dataset_id, value) VALUES ( '%d', '%d', '%s' )", $meta_id, $dataset_id, maybe_serialize($new_value) ) );
	
		// Textarea
		if ( 'textfield' == $formfield_type ) {
			$new_value = str_replace("\n", "", $new_value);
		if (strlen($new_value) > 150 )
			$new_value = substr($new_value, 0, 150)."...";
		}
			
		// Some special output formats
		if ( 'date' == $formfield_type )
			$new_value = mysql2date(get_option('date_format'), $_POST['new_value']);
		elseif ( 'image' == $formfield_type )
			$new_value = '<img class="projectmanager_image" src="'.$new_value.'" alt="'.__("Image", "projectmanager").'" />';
		elseif ( 'uri' == $formfield_type )
			$new_value = '<a class="projectmanager_url" href="http://'.$projectmanager->extractURL($new_value, 'url').'" target="_blank" title="'.$projectmanager->extractURL($new_value, 'title').'">'.$projectmanager->extractURL($new_value, 'title').'</a>';
		elseif ( 'email' == $formfield_type )
			$new_value = '<a href="mailto:'.$projectmanager->extractURL($new_value, 'url').'" class="projectmanager_email">'.$projectmanager->extractURL($new_value, 'title').'</a>';	
		elseif ( 'numeric' == $formfield_type ) {
			$new_value = apply_filters( 'projectmanager_numeric', $new_value );
		} elseif ( 'currency' == $formfield_type ) {
			$new_value = money_format('%i', $new_value);
			$new_value = apply_filters( 'projectmanager_currency', $new_value );
		} elseif ( 'checkbox' == $formfield_type || 'project' == $formfield_type ) {
			$list = '<ul class="'.$formfield_type.'" id="form_field_'.$meta_id.'">';
			foreach ( (array)$new_value AS $item ) {
				if ( 'project' == $formfield_type && is_numeric($item) ) {
					$item = $projectmanager->getDataset($item);
					if ( $_GET['page'] == 'projectmanager' )
						$url_pattern = '<a href="admin.php?page=projectmanager&subpage=dataset&edit='.$item->id.'&project_id='.$item->project_id.'">%s</a>';
					else
						$url_pattern = '<a href="admin.php?page=project-dataset_'.$item->project_id.'&edit='.$item->id.'&project_id='.$item->project_id.'">%s</a>';
					
					$item = sprintf($url_pattern, $item->name);
				}
				$list .= '<li>'.$item.'</li>';
			}
			$list .= '</ul>';
			$new_value = $list;
		}

		die( "
			jQuery('span#datafield" . $meta_id . "_" . $dataset_id . "').fadeOut('fast', function() {
			jQuery('a#thickboxlink" . $meta_id . "_" . $dataset_id . "').show();
			jQuery('span#datafield" . $meta_id . "_" . $dataset_id . "').html('" . $new_value . "').fadeIn('fast');
			ProjectManager.doneLoading('loading_".$meta_id."_".$dataset_id."');
			ProjectManager.reInit();
		});");
	}


	/**
	 * SACK response to manually set order of datasets
	 *
	 * @since 2.0
	 */
	function saveDatasetOrder() {
		global $wpdb, $projectmanager_loader;
		$order = $_POST['order'];
		$order = $projectmanager_loader->adminPanel->getOrder($order);
		foreach ( $order AS $order => $dataset_id ) {
			$wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->projectmanager_dataset} SET `order` = '%d' WHERE `id` = '%d'", $order, $dataset_id ) );
		}
	}


	/**
	 * SACK response to insert user data from database
	 *
	 * @since 2.5
	 */
	function insertWpUser() {
		$user_id = (int)$_POST['wp_user_id'];
		$user = new WP_User($user_id);
		$user = $user->data;

		die("
			document.getElementById('name').value = '".$user->display_name."';
			document.getElementById('user_id').value = '".$user_id."';
		");
	}


	/**
	 * get category dropdown
	 *
	 * @param none
	 * @since 2.7
	 */
	function getCategoryDropdown()
	{
		global $projectmanager;

		$project_id = (int)$_POST['project_id'];
		$formfield_id = (int)$_POST['formfield_id'];

		if ( !empty($project_id) ) {
			$p = $projectmanager->getProject($project_id);

			if ( !empty($p->category) && $p->category != -1 ) {
				$html = wp_dropdown_categories(array('hide_empty' => 0, 'name' => "form_field[".$formfield_id."][category]", 'orderby' => 'name', 'echo' => 0, 'show_option_none' => __('Select Categoy (Optional)', 'projectmanager'), 'child_of' => $p->category));
				$html = str_replace("\n", "", $html);
				die("jQuery('span#project_cat_".$formfield_id."').fadeOut('fast', function () {
					jQuery('span#project_cat_".$formfield_id."').html('".addslashes_gpc($html)."').fadeIn('fast');
				})");
			} else {
				die("jQuery('span#project_cat_".$formfield_id."').fadeOut('fast');");
			}
		} else {
			die("jQuery('span#project_cat_".$formfield_id."').fadeOut('fast');");
		}
	}


	/**
	 * toggle Formfield Options
	 *
	 * @since 2.7
	 */
	function toggleFormfieldOptions()
	{
		global $projectmanager_loader, $projectmanager;

		$project_id = (int)$_POST['project_id'];
		$formfield_id = (int)$_POST['formfield_id'];
		$formfield_type = (string)$_POST['formfield_type'];
		$options = (string)$_POST['options'];
		$options = explode("|", $options);

		$admin = $projectmanager_loader->getAdminPanel();

		$html = '';

		if ( 'project' == $formfield_type || 'checkbox' == $formfield_type || 'radio' == $formfield_type || 'select' == $formfield_type ) {
//			$html = '<div id="form_field_options_container'.$formfield_id.'" style="display: inline;">';
			$html = '<div id="form_field_options_div'.$formfield_id.'" style="overflow: auto; display: none;">';
//			$html .= '<form>';
			if ( 'project' == $formfield_type ) {
				$dims = array( 'width' => 300, 'height' => 100 );
				$thickbox_title = __( 'Choose Project to Link', 'projectmanager' );
				$icon = 'databases.png';
				
				$html .= '<div class="thickbox_content">';
				$html .= '<select size="1" id="form_field_project_'.$formfield_id.'">';
				$html .= '<option value="0">'.__( 'Choose Project', 'projectmanager' ).'</option>';
				foreach ( $projectmanager->getProjects() AS $p ) {
					if ( $p->id != $project_id )
						$html .= '<option value="'.$p->id.'">'.$p->title.'</option>';
				}
				$html .= '</select>';
				$html .= '<div class="buttonbar"><input type="button" value="'.__('Save').'" class="button-secondary" onclick="ProjectManager.saveProjectLink('.$formfield_id.');return false;" />&#160;<input type="button" value="'.__('Cancel').'" class="button" onclick="tb_remove();" /></div></div>';
			} elseif ( 'checkbox' == $formfield_type || 'radio' == $formfield_type || 'select' == $formfield_type ) {
				$dims = array( 'width' => 350, 'height' => 200 );
				$thickbox_title = __( 'Options', 'projectmanager' );
				$icon = 'application_list.png';

				$html .= '<div class="thickbox_content"><div class="">';
				$html .= '<ul id="form_field_options_'.$formfield_id.'">';
				foreach ( (array)$options AS $x => $item ) {
					$html .= '<li id="form_field_option_'.$formfield_id.'_'.$x.'"><input type="text" size="30" value="'.$item.'" name="form_field_option_'.$formfield_id.'" /><a class="image_link" href="#" onclick="return ProjectManager.removeFormFieldOption(\'form_field_option_'.$formfield_id.'_'.$x.'\', '.$formfield_id.');"><img src="../wp-content/plugins/projectmanager/admin/icons/trash.gif" alt="'.__( 'Delete', 'projectmanager' ).'" title="'.__( 'Delete Option', 'projectmanager' ).'" /></a></li>';
				}
				$html .= '</ul>';
				$html .= '</div>';

				$html .= '<p><a href="#" onClick="ProjectManager.addFormFieldOption('.$formfield_id.')">'.__( 'Add Option', 'projectmanager' ).'</a></p>';

				$html .= '<div class="buttonbar"><input type="button" value="'.__('Save').'" class="button-secondary" onclick="ProjectManager.ajaxSaveFormFieldOptions('.$formfield_id.');return false;" />&#160;<input type="button" value="'.__('Cancel').'" class="button" onclick="tb_remove();" /></div>';
				$html .= '</div>';
			}

//			$html .= '</form>';
			$html .= '</div>';
			$html .= '<span>&#160;<a href="#TB_inline&width='.$dims['width'].'&height='.$dims['height'].'&inlineId=form_field_options_div'.$formfield_id.'" style="display: inline;" id="options_link'.$formfield_id.'" class="thickbox" title="'.$thickbox_title.'"><img src="'.$admin->getIconURL($icon).'" alt="'.__('Options','projectmanager').'" class="middle" /></a></span>';
//			$html .= '</div>';
		}

		die("
			var type = '".$formfield_type."';
			new_element = document.createElement('div');
			new_element_id = 'form_field_options_container".$formfield_id."';
			new_element.id= new_element_id;
			new_element.style.display = 'inline';
			// Check if selected form type is selection, checkbox, or radio
			if ( type == 'project' || type == 'select' || type == 'checkbox' || type == 'radio' ) {
				if ( document.getElementById(new_element_id) ) {
					jQuery('div#form_field_options_container".$formfield_id."').fadeOut('fast');
				} else {
					if ( document.getElementById('form_field_optionos_box".$formfield_id."') ) {
						document.getElementById('form_field_options_box".$formfield_id."').appendChild(new_element);
					} else {
						alert('".__('An Error Occured. Please Save FormFields and set Options afterwards', 'projectmanager')."');
					}
				}
				jQuery('div#form_field_options_container".$formfield_id."').html(\"".addslashes_gpc($html)."\").fadeIn('fast');
				ProjectManager.reInit();
			} else {
				jQuery('div#form_field_options_container".$formfield_id."').fadeOut('fast');
				if (target_element = document.getElementById(new_element_id)) {;
					document.getElementById('form_field_options_box".$formfield_id."').removeChild(target_element);
				}
			}

		");
			//ProjectManager.doneLoading('loading_formfield_options_".$formfield_id."');
	}
}
